fix: Enhance IME Suggestions & UI — Bug Fixes, Refactors, and Performance Improvements

bhanu-dev82 opened the pull request:
<!---
Thank you for your pull request! 🚀
-->

### Contributor checklist

<!-- Please replace the empty checkboxes [] below with checked ones [x] accordingly. -->

-   [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch
-   [x] I have tested my code with the `./gradlew lintKotlin detekt test` command as directed in the [testing section of the contributing guide](https://github.com/scribe-org/Scribe-Android/blob/main/CONTRIBUTING.md#testing)

---

### Description

**Fixes:** #398 (broken gender suggestions); also addresses plural and case suggestion issues.  

This PR refactors and centralizes the word suggestion logic in `KeyHandler.kt` to ensure consistent behavior for German noun type, plural, and case annotation suggestions — especially after space, delete, and regular key presses.

---

### ✅ Key Changes

#### 🔁 Unified Suggestion Handling
- Added `processWordSuggestions(currentWord)` to encapsulate suggestion logic.
- Called after **space**, **delete**, and **character** key presses (in idle mode).

#### ⌨️ Space & Delete Logic
- `handleKeycodeSpace()` now commits a space, then triggers suggestions.
- `handleDeleteKey()` updates or clears suggestions after a delete.
- Skips suggestions in **command bar** context.

#### 🔤 Character Input
- `handleDefaultKey()` disables suggestions during command bar usage.

#### ♻️ Suggestion Clearing
- Inlined logic to reset noun/plural/case/emoji states in multiple handlers (enter, arrow, etc.).

#### 😊 Emoji Fallback
- If no noun/case suggestions found, emojis are suggested via `updateEmojiSuggestions()`.

---

### 🐛 Bug Fixes

- **Gender Suggestion Missing on Space:** Now correctly appears using `processWordSuggestions()`.
- **Incorrect "PL" Suggestion on Empty Field:** Fixed with null/empty check.
- **Inconsistent Suggestions:** Unified flow ensures reliable updates across keys.

---

### 🧪 Testing

- `"buch "` → correctly shows `"N"` (neuter)
- Deleting characters updates or clears suggestions appropriately
- Logs confirm expected flow through `processWordSuggestions()`

---

### 🔍 Note and Demo

This issue was approached with a test-driven and results-oriented development process. Community feedback and additional testing are encouraged to ensure comprehensive coverage and robustness across use cases.


https://github.com/user-attachments/assets/10027e67-0487-49c6-a946-f4fe281a78cd



---

Closes: #398

andrewtavis commented:
Thanks so much for this, @bhanu-dev82! Looking forward to the review as always as the quality of the information is so helpful 😊

andrewtavis commented:
Thanks for fixing the merge conflict and checking the functionality, @bhanu-dev82! Just checked the changes and all's working great in terms of the work for #398, but it looks like the change from #412 has been lost 🤔 Specifically for this we want the noun annotation to after commands. Here are two situations for the German keyboard that should function:

- [ ] If we translate the English word "book" on the German keyboard and get back "Buch", then we should see the neutral noun annotation (a green N should be in the picture below)
- [ ] Another note on the above is that we're not getting an extra space added in after the translated word
    - Right now we're getting back "Buch" from "book", but we should be getting "Buch " with a space after

![Screenshot_20250529_222405](https://github.com/user-attachments/assets/43c030bf-a91c-40e9-be6c-33e244245202)

- [ ] If we get the plural for the German word "Buch" on the German keyboard and get back "Bücher", then we should see the plural noun annotation (an orange PL should be in the picture below)
    - This command does correctly return an extra space after the output word :)

![Screenshot_20250529_222429](https://github.com/user-attachments/assets/a5948ae3-559b-4840-b757-19dfb8795727)

Please let us know if you have questions on the above! I'll let @angrezichatterbox provide feedback on the code specifics 😊

Thank you for being such a great contributor to this project! Your collaborator status was more than earned 😊 Looking forward to continuing to work with you! It really has been a pleasure :)

bhanu-dev82 commented:
yes i am just completing the fix for it, it will soon be pushed. Thankyou

andrewtavis commented:
Amazing, @bhanu-dev82! Thanks for getting to all of this so quickly! 😊

bhanu-dev82 commented:
i have fixed the errors:

* translate btn now add space after each word
* plural btn hve proper PL suggestions
* gender suggestions works
* multiple suggestions (PL/M, PL/F and more) works

I have made 2 new files namely SpaceKeyProcessor.kt and SuggestionHandler.kt   just to make detekt forgive me :stuck_out_tongue_winking_eye: 

here is a video showing the fix : 

Uploading Screen_Recording_20250530_024810_Keep Notes.mp4…

angrezichatterbox reviewed (commented):
Thanks for refactoring the codebase while fixing the bug @bhanu-dev82.

There is a slight issue in the implementation that causes the emoji suggestions to disappear when the command bar suggestion shows up. The actual behavior would be that both would appear simultaneously.

angrezichatterbox commented:
@andrewtavis I assume the intended behavior would be both to be shown simultaneously.

andrewtavis commented:
Assuming you mean noun annotation, @angrezichatterbox? Looking at the video, it looks like we're just getting the annotations like `PL` after commands, but yes, we should also get the emoji autosuggestios as well 😊

angrezichatterbox commented:
> Assuming you mean noun annotation, @angrezichatterbox? Looking at the most recent video, it looks like we're just getting the annotations like `PL` after commands, but yes, we should also get the emoji autosuggestios as well 😊

Yes, Currently when we type `buch` only the gender suggestion show up rather than the emoji suggestion also showing up.

andrewtavis commented:
Yes, it'd be important for us to show both :)

Are the colored `F`, `M`, `N`, etc called suggestions in the code? Maybe something I didn't see in review. Would be good if we had some consistency with the iOS app for this, with these being "annotations". We already have autosuggestions and emoji suggestions, so having these be known as something distinct would likely be good :) We could make an issue for the hackathon to rename related variables if this is the case and you're fine with the naming convention, @angrezichatterbox 😊

bhanu-dev82 commented:
> > Assuming you mean noun annotation, @angrezichatterbox? Looking at the most recent video, it looks like we're just getting the annotations like `PL` after commands, but yes, we should also get the emoji autosuggestios as well 😊
> 
> Yes, Currently when we type `buch` only the gender suggestion show up rather than the emoji suggestion also showing up.

Apologies for the delayed response — I've been occupied with exams. The emoji issue will be addressed through the fix for #404, as it includes handling for emoji-related cases. If possible, could we proceed with this approach? if not i will fix it and push in the same PR. i will be working on #404 today.

andrewtavis commented:
I guess if you're fine with it @bhanu-dev82, just send along the changes to this branch and we'll finalize both issues in one PR? That way we can check the functionalities together 😊

bhanu-dev82 commented:
Due to a delay on my part in updating this pull request, merge conflicts have now arisen. Resolving them would require significant time, so instead, I will incorporate the necessary changes into an updated version of the code to avoid these conflicts. Therefore, I’m closing this PR.

bhanu-dev82 commented:
The recent merge was intended solely to integrate the latest changes. A major update is in progress, which significantly improves performance. Specifically, the complexity of generating suggestions has been optimized from O(n) to O(1), resulting in a much faster keyboard and an overall smoother user experience.

bhanu-dev82 commented:
I hate to have 10s of 20 s problem in one pull request but because i am fixing 3 issues, this pull requst is going to be huge. I have also edited files like database helper.

bhanu-dev82 commented:
**More improvements are on the way to streamline database handling and suggestion logic.**

The latest commit includes the following fixes and enhancements:

    Resolved issues with gender and plural suggestions.

    Fixed spacing issues with the translate buttons.

    Addressed a visual rendering error across all IMEs by applying proper inheritance handling.

andrewtavis commented:
> I hate to have 10s of 20 s problem in one pull request but because i am fixing 3 issues, this pull requst is going to be huge. I have also edited files like database helper. 

Totally appreciate this, @bhanu-dev82. Definitely was not the plan at the outset that this would expand to where it is now. Appreciate you pushing through this one, and let's definitely keep this in mind for another time that we choose to pull other issues into an already open PR. I was also keeping capacity for @angrezichatterbox and I in mind and wanted to make sure that you could continue to work on things at a time when we were both busy, but maybe it would have been best to do two separate PRs :)

Would you like to bring this in and make separate issues for the rest? Let us know when you think the work here is ready for a final review! 😊

bhanu-dev82 commented:
> > I hate to have 10s of 20 s problem in one pull request but because i am fixing 3 issues, this pull requst is going to be huge. I have also edited files like database helper.
> 
> Totally appreciate this, @bhanu-dev82. Definitely was not the plan at the outset that this would expand to where it is now. Appreciate you pushing through this one, and let's definitely keep this in mind for another time that we choose to pull other issues into an already open PR. I was also keeping capacity for @angrezichatterbox and I in mind and wanted to make sure that you could continue to work on things at a time when we were both busy, but maybe it would have been best to do two separate PRs :)
> 
> Would you like to bring this in and make separate issues for the rest? Let us know when you think the work here is ready for a final review! 😊

Thanks for the clarification, and I totally understand where you're coming from. Just to clarify on my end—I wasn’t complaining, but rather trying to make things smoother for you both, since reviewing such a large PR can be quite a task. If I were in your position, I know it would’ve been a bit of a brain-burner!

That said, most of the core issues have now been addressed, and I’ll be pushing two final commits shortly to wrap things up. After that, it should be ready for a final review. I agree that separating concerns across multiple PRs would be a better approach moving forward—definitely something I’ll keep in mind.

Thanks again for your support throughout! 😊

andrewtavis commented:
Wasn't taking it as a complaint, @bhanu-dev82 :) Just trying to recognize that the process here has been a bit weird this time around and voicing ways that we can do a bit better moving forward 😊 Was also a bit of self reflection to remember to keep things more scoped in the future :)

bhanu-dev82 commented:
with this i have finished the required task in this pull request. the code has been tasted for functionality, performance is concern as ui is getting refreshed for every key stroke, with is a old issue but this can be addressed in another pull request.

andrewtavis commented:
@bhanu-dev82, looking at it the colors are not right on main as well, so this is coming in from #419 odds are. I'll make an issue now. If you could fix make it so the emoji separator only appears with the emojis that'd be great, but let's not worry about the colors here! 😊

bhanu-dev82 commented:
**With these changes, the purpose of this pull request is now fulfilled.**

This update includes a comprehensive refactoring of both **data handling** and **UI updates**, which are the core components of the affected functionality. I have taken extra care to test and verify the changes thoroughly to minimize the risk of introducing new bugs. However, more in-depth testing might still reveal edge-case issues, particularly related to UI behavior.

### Key Improvements:

* **Gender suggestions:** Enhanced logic for better accuracy.
* **Plural suggestions:** Fixed a bug affecting correct rendering.
* **IME-specific issues:** Resolved crashes in Russian and Swedish input methods during translation.
* **UI Refactor:** Introduced a new layout file (`input_method_view.xml`) to better structure the IME interface.
* **Performance:** Improved suggestion retrieval by replacing list searches with hash-based lookup and optimized UI update routines.

While the implementation has been carefully validated, please pay close attention to the new UI layout in testing, as that area is most susceptible to unforeseen issues.

### Screenshot for Gender suggestions
![Screenshot_20250615_130908_Keep Notes](https://github.com/user-attachments/assets/611bc804-a058-46aa-915f-5b99f61e55e8)
![Screenshot_20250615_130921_Keep Notes](https://github.com/user-attachments/assets/d568c042-b9a4-449a-ad64-9a7a38cce21b)
![Screenshot_20250615_130930_Keep Notes](https://github.com/user-attachments/assets/9eff6bbf-7bfc-4f04-8ab3-8b24c69451ea)
![Screenshot_20250615_130858_Keep Notes](https://github.com/user-attachments/assets/85adb19f-2a63-49ad-b592-81fd48c8280e)
![Screenshot_20250615_130850_Keep Notes](https://github.com/user-attachments/assets/1f487b39-e01f-49af-adb9-70d244f30e8c)

andrewtavis commented:
We're very very close here, @bhanu-dev82 :) Checking it out now, and there are just some very minor changes that are needed. I'll order them based on should be in this PR and ones that could be other issues:

Should be in this PR:
- There's a coloration to the suggestion buttons for autosuggestions now (see image below)
    - Ideally we'd not have this (I'm on Pixel 9 and API 35 for the testing)

![Screenshot_20250615_144440](https://github.com/user-attachments/assets/864ecece-1c03-47aa-b960-eb0e6225cc47)

Can be new issues:
- The style of the separator for emojis isn't the same as the ones for suggestions
    - I like the one for emojis better, so it'd be great if we could make the autosuggestion ones look like that one

![Screenshot_20250615_144650](https://github.com/user-attachments/assets/5c4db0ea-16dd-402d-a92e-99472fdbfbbd)

- Ideally the full annotations would only show after the user has pressed space (this should be the trigger for all annotations)
    - The user types "Buch" for "book", and at that point we'd see the emoji suggestions and two autosuggestion fields
    - Only when the user presses space for "Buch " would we then see the `N` annotation
    - We'll have another functionality added later that will do annotations as the user types, but this issue isn't made yet (I'll make it now)

I'd say we'd be good to merge this once the suggestion buttons are colored appropriately (i.e. not colored) 😊

bhanu-dev82 commented:
The way elements render on device and virtual machine differs a lot there are ways to handle the tint one is through code other is through XML, the XML is non dynamic. so i implemented both but now we will only use dynamic implementation from code. so the error 1 is fixed

about error 2 the separators help create a visual gap between the button which appears after clicking the scribe key, i have not taken into account the dp to px conversion that caused the error hopeful it is fixed now i had just a random guess and it fixed it. 

about the third one it simple to fix to have suggestions only after user uses the space key.

i will do this changes quickly

andrewtavis commented:
Thanks so much for getting to all of this, @bhanu-dev82! Really appreciate your dedication in this PR 🙏🙏

andrewtavis commented:
Commit above is purely some comment work to standardize it a bit and marking out dependencies, which did lead to realizing that `androidTestImplementation("androidx.compose.ui:ui-test-junit4:1.8.2")` was included in `app/build.gradle.kts` twice :)

bhanu-dev82 commented:
Thank you @andrewtavis, I am curious about how one-line comments should be written (those starting with //).

From my observation of the latest push:
* If comments start on their own line, they have a capital first letter and end with a period.
* If they are on the same line as code, they use lowercase letters. am i right?

Apologies for asking something that looks like basic English sentence writing.

andrewtavis commented:
The conventions you observed are exactly correct, @bhanu-dev82 :) And no need to apologize whatsoever. It's really not too important, but just something I like to address when I see it. I'm not sure what Kotlin's style guide says on this, but this is Python's and I tend to use that for the projects I work on 😊

andrewtavis commented:
@angrezichatterbox, at your convenience, could you give us a review of the code here? From there we can merge this in and support @DeleMike and @catreedle with any merge conflicts to their PRs 😊

angrezichatterbox commented:
I am doing some additional chores that I have noticed within the same PR. As I am already reviewing the codebase quite a bit.

angrezichatterbox commented:
@andrewtavis It is ready for review. I have tried seeing how the refactoring would be and refactored some of the components in PR. However for the complete refactoring we could have a separate issue as being discussed in #426.

andrewtavis commented:
@angrezichatterbox, not sure if it's just on my end, but there seem to be issued introduced and some of the changes that @bhanu-dev82 sent along seem to be reverted? See the screenshot below:

![Screenshot_20250621_001626](https://github.com/user-attachments/assets/b9ab5387-b154-494d-b1bf-3343167aa3a6)

What I'm seeing is:
- We're not getting annotations at all except for the `PL` when the keyboard starts
- The annotations are Scribe blue again instead of their colors
- The third autocompletion/suggestion isn't present when there are no emojis being shown

Can you take a look at the above and see if you can figure it out? If it's overly complex, then we could also revert back to 
4f33e50 and merge and then further changes can be done via new issues :) Let us know how you'd like to proceed, and if you do want to try to fix it, then feel free to get started 😊

andrewtavis commented:
Project isn't building on my end, even after clearing the caches. Are there other steps that I could do to get it working?

andrewtavis commented:
```
Execution failed for task ':app:processCoreDebugResources'.
> Could not resolve all files for configuration ':app:coreDebugRuntimeClasspath'.
   > Failed to transform biometric-1.2.0-alpha05.aar (androidx.biometric:biometric:1.2.0-alpha05) to match attributes {artifactType=android-compiled-dependencies-resources, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.libraryelements=aar, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Execution failed for AarResourcesCompilerTransform: PATH/.gradle/caches/8.11.1/transforms/7dd7686cc8712ff5bf13738323e503b0/transformed/biometric-1.2.0-alpha05.
         > AAPT2 aapt2-8.10.1-12782657-osx Daemon #0: Daemon startup failed
           This should not happen under normal circumstances, please file an issue if it does.
   > Failed to transform appcompat-1.7.1.aar (androidx.appcompat:appcompat:1.7.1) to match attributes {artifactType=android-compiled-dependencies-resources, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.libraryelements=aar, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Execution failed for AarResourcesCompilerTransform: PATH/.gradle/caches/8.11.1/transforms/d7fd5fcd804e9647c3a8b284a003996a/transformed/appcompat-1.7.1.
         > AAPT2 aapt2-8.10.1-12782657-osx Daemon #1: Daemon startup failed
           This should not happen under normal circumstances, please file an issue if it does.
   > Failed to transform core-1.16.0.aar (androidx.core:core:1.16.0) to match attributes {artifactType=android-compiled-dependencies-resources, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.libraryelements=aar, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Execution failed for AarResourcesCompilerTransform: PATH/.gradle/caches/8.11.1/transforms/58c640cf083e1536287e06acc04912f9/transformed/core-1.16.0.
         > AAPT2 aapt2-8.10.1-12782657-osx Daemon #2: Daemon startup failed
           This should not happen under normal circumstances, please file an issue if it does.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
BUILD FAILED in 21s
30 actionable tasks: 29 executed, 1 up-to-date
```

angrezichatterbox commented:
Could you try running `./gradlew clean` and `./gradlew build` It build fine for me. The errors seem to be from cache. I could revert back the `build.gradle.kts`

angrezichatterbox commented:
Could you check now. Or else I would revert back the entire `build.gradle.kts`. I have downgraded the ksp to support kotlin 2.0.0

andrewtavis commented:
`./gradlew build` failed with the following:

```
> Task :app:packageCoreRelease FAILED
[Incubating] Problems report is available at: file:///Users/andrewtavis/coding/scribe-org/Scribe-Android/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.
```

The info that's in the problem report is:

```
    - [warn]  The Configuration.fileCollection(Spec) method has been deprecated. (10)
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
            - [warn]  The Configuration.fileCollection(Spec) method has been deprecated.
```

The build works now within Android Studio, then just the app loads, I can install a keyboard within the settings, but I can't use the keyboard 🤔

angrezichatterbox commented:
For API 36 emulator none of the external downloaded keyboard works. Like I tried the fossify keyboard we can't switch to it as well.

andrewtavis reviewed (approved):
All's working with Pixel 9 and API 35 ✨🎉 Amazing work here, @bhanu-dev82! Really commendable how much we've progressed in this single PR 😊 Thanks also for the review and the refactoring/updates, @angrezichatterbox! 🙏

So glad to have this one in :) Let's focus on bringing in #395 and #420 now. You'd be welcome to start reviewing PRs too, @bhanu-dev82! Please let us know if this would be something you'd have interest in. As it suits you 😊

andrewtavis commented:
There's enough work in here that I'm not going to squash so it's a bit more reflective of the effort that went into this. Let's also discuss what's up with API 36... Likely is just something that will come in a forthcoming update.

DeleMike commented:
Great work here! @bhanu-dev82 

Really inspiring ⭐🙌🏾✨
